### ISSUE TITLE
Robyn_output() is not generating any of the CSVs or One Pagers for the selected clusters

### ISSUE THREAD

**Issue by CJ2407** [facebookexperimental/Robyn#603]
## Robyn_Output()

My model is completing with no errors. `robyn_outputs()` is completing with no errors. But in the output folder, none of the CSVs or One pagers are getting generated. Please assist. 

![image](https://user-images.githubusercontent.com/91983160/212170978-dd2c89da-05f9-4fcd-9311-7aa6bcc01744.png)

![image](https://user-images.githubusercontent.com/91983160/212171075-0bed0ec9-88ea-495c-8238-0666fde5fda2.png)

I am using Robyn 3.9.0
And R version is 4.2.2

**Comment by laresbernardo** [comment]
@Amyhaoming can you please help @CJ2407 debug? Does `robyn_object` directory exist? If that screenshot is showing everything on the output prints, it's behaving exactly as if `export = FALSE`. When `export = TRUE`, you should have a tail of messages similar to these:
```
(...)
>> Pareto-Front: 8 [10 models]
 00:00:02 [===========================================================] 100% | 10                                        
>>> Calculating clusters for model selection using Pareto fronts...
>> Auto selected k = 5 (clusters) based on minimum WSS variance of 5%
>>> Collecting 102 pareto-optimum results into: /Users/bernardolares/Desktop/Robyn_202301121524_init/
>> Exporting general plots into directory...
>> Exporting pareto results as CSVs into directory...
>>> Exporting pareto one-pagers into directory...
>> Generating only cluster results one-pagers (5)...
>> Plotting 5 selected models on 1 core (MacOS fallback)...
  |===========================================================================================================| 100%
>> Exported model inputs as /Users/bernardolares/Desktop/Robyn_202301121524_init/RobynModel-inputs.json
```

**Comment by CJ2407** [comment]
Yes, robyn_object is setup as "robyn_object <- "/mnt/data/cherry/"" I am running models in Rstudio setup in Gcloud VM instance. But, yeah messages on my end are ending at below line -

>> Auto selected k = 7 (clusters) based on minimum WSS variance of 5%

But a folder is getting created with only the following outputs - 

![image](https://user-images.githubusercontent.com/91983160/212174510-4a6f07bf-cdea-4650-9838-0aea38dfd062.png)

I am not understanding it why. When I run model on Rstudio on my local machine. I don't face the same problem. But I had to run it on VM to be able to share my work with my peers and take advantage of bigger memory machine

**Comment by laresbernardo** [comment]
> but now cluster numbers are also appearing with missing digits towards the end

That's actually a problem with the size of the window you're using to visualize the plot. Try and stretch it or change the font size and you'll see they're there.

> When I run model on Rstudio on my local machine. I don't face the same problem.

Both have the same Robyn version running?

> top models selected in each cluster seeing the cluster names from the below plot

Check `OutputCollect$clusters$data[OutputCollect$clusters$data$top_sol == TRUE, ]`. It'll show you a data frame with these top models and their IDs.

It'll be hard for us to debug because we don't have a VM (not to mention with same specifics) to replicate this problem, which is the first time I see. If you get to understand what's happening, please share with us.

**Comment by CJ2407** [comment]
Is there a way similar to robyn_onepagers() to get CSV outputs saved even if my robyn_output is currently working as export = FALSE?

**Comment by laresbernardo** [comment]
You can try `Robyn:::robyn_csv` unexported function.

**Comment by CJ2407** [comment]
Hey! robyn_csv() is giving me the following error - 
Error in check_class("robyn_outputs", OutputCollect) : 
  argument "OutputCollect" is missing, with no default

whereas robyn_onepagers(InputCollect5, OutputCollect, export = TRUE, select_model = "3_409_5") is working fine. Any solution?

**Comment by laresbernardo** [comment]
You must have `InputCollect` and `OutputCollect` as available objects to be able to run `robyn_csv(InputCollect, OutputCollect, "all")`. Please, read documentation: `?robyn_csv`.

**Comment by CJ2407** [comment]
I ran it like this - robyn_csv(OutputCollect, csv_out = "pareto", export = TRUE) and I do have OutputCollect in my global environment of R studio. Sorry, but I didn't understand what to do with OutputCollect besides that I just ran the following code before trying to run the above robyn_csv code - 
OutputCollect <- robyn_outputs(
  InputCollect5, OutputModels5,
  pareto_fronts = "auto", # automatically pick how many pareto-fronts to fill min_candidates
  # min_candidates = 100, # top pareto models for clustering. Default to 100
  # calibration_constraint = 0.1, # range c(0.01, 0.1) & default at 0.1
  csv_out = "pareto", # "pareto", "all", or NULL (for none)
  clusters = TRUE, # Set to TRUE to cluster similar models by ROAS. See ?robyn_clusters
  plot_pareto = TRUE, # Set to FALSE to deactivate plotting and saving model one-pagers
  plot_folder = robyn_object, # path for plots export
  export = TRUE # this will create files locally
)

**Comment by ddempseyr** [comment]
Is there any update on this issue? 

We are having the same issue after updating Robyn from v3.6.3 to v3.9.0. The CSV files are successfully created when locally running robyn_outputs on R Studio but CSV files are not created when run in a Docker container (UBUNTU_VER=20.04, OS_TYPE=x86_64).

**Comment by ddempseyr** [comment]
I'm not sure what is causing the issue but I found a discrepancy in the logs from robyn_outputs on RStudio vs. the logs from  Docker. Robyn_outputs returns an error on Docker stating "Error in ts_validation(OutputModels, quiet = TRUE, ...): object 'OutputModels' not found" which is odd because the code base is the same and ts_validation is set to FALSE in robyn_run. 


RStudio output
`
>>> Running Pareto calculations for 500 models on auto fronts...
>> Automatically selected 8 Pareto-fronts to contain at least 100 pareto-optimal models (103)
>> Pareto-Front: 1 [9 models]
 00:00:02 [=========================================================================================] 100% | 9                                                                       
>> Pareto-Front: 2 [10 models]
 00:00:02 [=========================================================================================] 100% | 10                                                                      
>> Pareto-Front: 3 [10 models]
 00:00:02 [=========================================================================================] 100% | 10                                                                      
>> Pareto-Front: 4 [13 models]
 00:00:05 [=========================================================================================] 100% | 13                                                                      
>> Pareto-Front: 5 [12 models]
 00:00:04 [=========================================================================================] 100% | 12                                                                      
>> Pareto-Front: 6 [13 models]
 00:00:03 [=========================================================================================] 100% | 13                                                                      
>> Pareto-Front: 7 [16 models]
 00:00:04 [=========================================================================================] 100% | 16                                                                      
>> Pareto-Front: 8 [20 models]
 00:00:04 [=========================================================================================] 100% | 20                                                                      
>>> Calculating clusters for model selection using Pareto fronts...
>> Auto selected k = 9 (clusters) based on minimum WSS variance of 5%
>>> Collecting 103 pareto-optimum results into: /Users/ddempsey/Desktop/Robyn_202302091122_init/
>> Exporting general plots into directory...
>> Exporting pareto results as CSVs into directory...
>>> Exporting pareto one-pagers into directory...
>> Generating only cluster results one-pagers (9)...
>> Plotting 9 selected models on 1 core (MacOS fallback)...
  |======================================================================================================================================================================| 100%
>> Exported model inputs as /Users/ddempsey/Desktop/Robyn_202302091122_init/RobynModel-inputs.json
`

Docker output
`
or67j8c5c8-algo-1-pjwo1 | >>> Running Pareto calculations for 500 models on auto fronts...
or67j8c5c8-algo-1-pjwo1 | >> Automatically selected 8 Pareto-fronts to contain at least 100 pareto-optimal models (113)
or67j8c5c8-algo-1-pjwo1 | >> Pareto-Front: 1 [10 models]
 00:00:04 [=========================================] 100% | 10                      
or67j8c5c8-algo-1-pjwo1 | >> Pareto-Front: 2 [9 models]
 00:00:03 [=========================================] 100% | 9                       
or67j8c5c8-algo-1-pjwo1 | >> Pareto-Front: 3 [16 models]
 00:00:05 [=========================================] 100% | 16                      
or67j8c5c8-algo-1-pjwo1 | >> Pareto-Front: 4 [14 models]
 00:00:05 [=========================================] 100% | 14                      
or67j8c5c8-algo-1-pjwo1 | >> Pareto-Front: 5 [10 models]
 00:00:04 [=========================================] 100% | 10                      
or67j8c5c8-algo-1-pjwo1 | >> Pareto-Front: 6 [13 models]
 00:00:05 [=========================================] 100% | 13                      
or67j8c5c8-algo-1-pjwo1 | >> Pareto-Front: 7 [20 models]
 00:00:08 [=========================================] 100% | 20                      
or67j8c5c8-algo-1-pjwo1 | >> Pareto-Front: 8 [21 models]
 00:00:08 [=========================================] 100% | 21                      
or67j8c5c8-algo-1-pjwo1 | >>> Calculating clusters for model selection using Pareto fronts...
or67j8c5c8-algo-1-pjwo1 | >> Auto selected k = 4 (clusters) based on minimum WSS variance of 5%
or67j8c5c8-algo-1-pjwo1 | >>> Collecting 113 pareto-optimum results into: /root/output/Robyn_202302091615_init/
or67j8c5c8-algo-1-pjwo1 | >> Exporting general plots into directory...
or67j8c5c8-algo-1-pjwo1 | Failed exporting results, but returned model results anyways:
or67j8c5c8-algo-1-pjwo1 |  Error in ts_validation(OutputModels, quiet = TRUE, ...): object 'OutputModels' not found
`

**Comment by laresbernardo** [comment]
Hi @ddempseyr 
Looking at Docker output, [that issue](https://github.com/facebookexperimental/Robyn/issues/596) was already fixed some weeks ago. Can you please update Docker's Robyn version to latest CRAN or dev version and retry please?

**Comment by ddempseyr** [comment]
Hi @laresbernardo I rebuilt the image with the latest version (3.9.0) from CRAN and attempted to run it within the docker container and am still seeing the same error. This error does not exist when we downgraded down to version 3.8.2. 

```
exi1qzthr4-algo-1-c80zb | [1] "Collecting Robyn outputs..."
exi1qzthr4-algo-1-c80zb | >>> Running Pareto calculations for 10000 models on 1 front...
exi1qzthr4-algo-1-c80zb | >> Pareto-Front: 1 [16 models]
 00:00:06 [=========================================] 100% | 16                      
exi1qzthr4-algo-1-c80zb | >>> Calculating clusters for model selection using Pareto fronts...
exi1qzthr4-algo-1-c80zb | >> Auto selected k = 9 (clusters) based on minimum WSS variance of 5%
exi1qzthr4-algo-1-c80zb | >>> Collecting 16 pareto-optimum results into: ./output/Robyn_202302222042_init/
exi1qzthr4-algo-1-c80zb | >> Exporting general plots into directory...
exi1qzthr4-algo-1-c80zb | Failed exporting results, but returned model results anyways:
exi1qzthr4-algo-1-c80zb |  Error in ts_validation(OutputModels, quiet = TRUE, ...): object 'OutputModels' not found
```

**Comment by laresbernardo** [comment]
Hi @ddempseyr are you 100% sure you've updated Robyn's version? That really shouldn't happen with latest versions as I mentioned given that was a known issue and was fixed before landing CRAN's latest version. Also, can you try running latest dev version that should be 3.9.1.9000? Print the version after `[1] "Collecting Robyn outputs..."`

**Comment by ddempseyr** [comment]
@laresbernardo Unfortunately, I am confident that the docker image was rebuilt with v3.9.0. I printed the package version to the console to be sure. 

Does the Github release reflect the files on CRAN for v3.9.0? If so then this bug fix may not have made it into v.3.9.0 because, based on the release notes, it seems like v3.9.0 was released back in December and the fix wasn't merged until Jan 20th (https://github.com/facebookexperimental/Robyn/pull/607/commits). 

In regards to running the dev version, I'd be happy to run a test. I'm a little swamped right now so it might take me a day or two to get back to you.

Thanks for your help!

**Comment by ddempseyr** [comment]
@laresbernardo I just noticed that our Docker container is installing 3.9.0 from github, not CRAN, and there could be differences in the code base since 3.9.0 was released on github back in December 2020 while 3.9.0 was published on CRAN in February 2023. 

I believe that we simply need to install from CRAN to get the latest bug fixes for 3.9.0. I'm running a test now to confirm. I will keep you posted.
